assets/src/adaptivity/rules-engine.ts

```diff
@@ -347,33 +350,55 @@ export const getReferencedKeysInConditions = (conditions: any) => {
       any: boolean;
       fact: string;
       value: string | number | boolean | unknown;
+      target?: string;
     }) => {
       // the fact *must* be a reference to a key we need
-      if (condition.fact) {
+      if (!checkExpressionInTarget && condition.fact) {
         references.add(condition.fact);
       }
       // the value *might* contain a reference to a key we need
       if (typeof condition.value === 'string') {
-        extractUniqueVariablesFromText(condition.value).forEach((v) => references.add(v));
+        const variablesInCondition = checkExpressionInTarget
```

Comment:
To improve the performance, when a lesson is opened in authoring, we generate a list of variables (Init state, Rules and SIM - > Config Data) that contains expression and needs evaluation. We stored them in conditionsNeedEvaluation in activity.content.custom.conditionsNeedEvaluation.